[Compose 페이먼츠 2단계] 페토 미션 제출합니다

chanho0908 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
#### 안녕하세요 오이 ! 🥒
1단계에 이어 2단계도 잘 부탁드립니다 👏 주말에 리뷰 요청드린 관계로 월요일 자정까지 리뷰해주셔도 괜찮습니다 !! 🙂

1단계에서 스스로 느끼기에 테스트가 많이 부족하다고 느꼈어요. TDD 미션을 진행하는 크루들을 보며 많은 자극을 받아 이번 단계는 TDD로 진행해봤어요 !!  😵‍💫

또한 결제 카드 컴포저블을 최대한 재활용 하기 위한 방법들을 고민해봤습니다 ! 그래서 카드를 세 분류로 정의해 봤어요 
+ EMPTY : 카드를 추가하기 위한 빈 형태의 카드
+ PENDING : 카드 추가 화면의 카드
+ REGISTERED : 추가 완료된 카드
```
enum class CardType(
    val parentAlignment: Alignment,
    val backgroundColor: Color,
) {
    EMPTY(Alignment.Center, GrayE5),
    PENDING(Alignment.CenterStart, Black33),
    REGISTERED(Alignment.CenterStart, Black33)
}
```

그리고 이 enum class들은 최소한의 공통 요소들만 가지도록 설계했어요. 이로 인해 Modifier와 카드 내부 형태를 구현할 컴포넌트는 외부에서 주입받도록 설계했습니다
```
@Composable
fun PaymentCard(
    cardType: CardType,
    content: @Composable () -> Unit,
    modifier: Modifier = Modifier,
    onClick: (CardType) -> Unit = {},
) 
```

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

https://github.com/user-attachments/assets/19e64bb5-1709-463f-ba85-ba8e4aadc9e3

chanho0908 committed:
docs: 2단계 기능 요구사항 추가

chanho0908 committed:
feat: 카드 목록 화면 타이틀 추가

chanho0908 committed:
feat: 카드 없을 때 새로운 카드를 등록해주세요 텍스트 추가

chanho0908 committed:
feat: 카드 도메인 모델 추가

chanho0908 committed:
feat: 등록된 카드 없는 경우 기본 카드 이미지 표시

chanho0908 committed:
refactor: CardNumberVisualTransformation을 외부에서 주입받도록 수정

chanho0908 committed:
feat: 카드 칩 컴포저블 분리

chanho0908 committed:
feat: 카드 번호 포맷팅 관련 상수화 및 `CardNumberTextField` 의존성 분리

chanho0908 committed:
feat: ExpireDateVisualTransformation 외부 주입으로 수정

chanho0908 committed:
feat: 카드 도메인에 카드 번호 및 만료일 포맷팅 기능 추가

chanho0908 committed:
feat: 등록된 카드 컴포넌트 추가

chanho0908 committed:
feat: 카드 종류에 따라 카드 컴포넌트 스타일 변경 기능 추가

chanho0908 committed:
feat: 등록된 카드가 1장 초과면 해당 카드들의 정보만 보여주는 기능 구현

chanho0908 committed:
feat: 카드 목록 상단 툴바 추가

chanho0908 committed:
feat: `kotlin-parcelize` 플러그인 추가

chanho0908 committed:
feat: 직렬화 카드 모델 추가

chanho0908 committed:
feat: Intent 확장 함수 `getParcelableCompat` 추가.

chanho0908 committed:
feat: 카드 추가 기능 구현

chanho0908 committed:
feat: 카드 추가 시 토스트 메시지 표시

chanho0908 committed:
refactor: 카드 목록 화면 툴바 컴포저블 수정

chanho0908 committed:
test: 도메인 카드 테스트 JUnit4로 테스트로 수정

chanho0908 commented on app/src/main/java/woowacourse/payments/cards/CardsActivity.kt:
```diff
@@ -0,0 +1,86 @@
+        super.onCreate(savedInstanceState)
+        enableEdgeToEdge()
+        setContent {
+            AndroidpaymentsTheme {
+                val cards = remember { mutableStateListOf<Card>() }
+                var showToast by remember { mutableStateOf(false) }
+
```
최초엔 카드가 없단다 ...

chanho0908 commented on app/src/main/java/woowacourse/payments/cards/CardsActivity.kt:
```diff
@@ -0,0 +1,86 @@
+                        },
+                        Modifier
+                            .padding(innerPadding)
+                    )
```
매번 이런 이벤트가 늘어날 때 마다 초기값으로 복원하는 로직이 불필요한 코드를 늘리는 것 같아
이를 최적화한 패턴이 있을까 ?

chanho0908 committed:
refactor: 카드 추가 버튼 TextButton으로 변경

chanho0908 committed:
feat: 일회성 이벤트 처리를 위한 Event 클래스 추가 및 적용

chanho0908 commented on app/src/main/java/woowacourse/payments/component/CardNumberTextField.kt:
```diff
@@ -10,22 +10,21 @@ import androidx.compose.ui.Modifier
 fun CardNumberTextField(
     maxLength: Int,
     cardNumber: String,
+    visualTransformation: VisualTransformation,
```
오이의 리뷰를 적용해 리컴포지션 되어도 매번 인스턴스가 생성되지 않도록 수정했습니다 :)

cucumber99 replied:
제가 생각한 해결 방법은 해당 객체를 `remember`로 선언하는 것이었는데,
이렇게 해도 괜찮을 것 같네요!

chanho0908 commented on app/src/main/java/woowacourse/payments/component/PaymentCard.kt:
```diff
@@ -1,46 +1,87 @@
 @Composable
 fun PaymentCard(
+    cardType: CardType,
+    content: @Composable () -> Unit,
```
이렇게 컴포저블을 인자로 받아 동적으로 컴포넌트를 그리는 것을 컴포즈에선 Slot Pattern 이라고 하더군요 
[참조](https://wisemuji.medium.com/jetpack-compose-ui-%EC%A1%B0%ED%95%A9-composition-%ED%95%98%EA%B8%B0-%EC%8B%AC%ED%99%94-33910e8f09df)(어라... 익숙한 미디엄이...?!)

cucumber99 replied:
샤라웃 투 레아

chanho0908 replied:
@woowahan-leah 레아의 글 잘 봤습니다 👍

woowahan-leah replied:
^^
어쩐지 clap 수가 늘었더라...

chanho0908 commented on app/src/main/java/woowacourse/payments/component/PaymentsToolbar.kt:
```diff
@@ -0,0 +1,60 @@
+    addButtonVisible: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    CenterAlignedTopAppBar(
```
자동으로 타이틀을 중앙 정렬해주는 컴포넌트입니다 토끼가 알려줘서 썼는데 신기방기 ,,,

chanho0908 commented on app/src/androidTest/java/woowacourse/payments/cards/CardsScreenTest.kt:
```diff
@@ -0,0 +1,134 @@
+        )
+
+        // when
+        composeTestRule.setContent {
+            CardsScreen(cards, CardScreenUiEvent.Idle, {})
+        }
```
이전 리뷰에서 레아가 컴포즈룰을 위한 컴포넌트 세팅하는 코드에 대해서 명확하게 알고 사용하는 것인지 리뷰주셨는데,

CardsScreen을 테스트하기 편하게 만들기 위해 상태를 외부에서 주입 받도록 만들었습니다. 때문에 
given : 특정 카드 목록이 있을 때
when : 이 카드 목록을 화면에 전달하면
then : 어떠한 결과가 나온다

라는 의도로 여길 when으로 생각했어요 ! 이하 동일합니다

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/cards/CardsScreenTest.kt:
```diff
@@ -0,0 +1,134 @@
+            .assertIsDisplayed()
+    }
+
+    @Test
+    fun `빈_카드_클릭_시_onClickCard가_호출된다`() {
```
여기도 given/when/then 주석을 추가하면 좋을 것 같아요~

chanho0908 replied:
https://github.com/woowacourse/android-payments/pull/28/commits/27e58c58fc0da295fc521bec657ba93d93b0c308

리뷰 적용 완~

cucumber99 commented on app/src/main/java/woowacourse/payments/MainActivity.kt:
```diff
@@ -17,18 +26,32 @@ class MainActivity : ComponentActivity() {
+                                val intent =
+                                    CardsActivity.newIntent(this, card.toSerializationCard())
+                                setResult(RESULT_OK, intent)
+                                finish()
```
메소드로 분리하면 가독성이 더 좋아지지 않을까요?

chanho0908 replied:
[628aed8](https://github.com/woowacourse/android-payments/pull/28/commits/628aed872ecd02cd15f5412d2692be2f644411b7) [9cd646d](https://github.com/woowacourse/android-payments/pull/28/commits/9cd646d4a42a1b7686b3c62da57604674020c2e9)

해당 리뷰를 보고 다시 생각해보니, 지금 Activiy가 UI + 상태 관리 + ActivityResult 처리까지 다 가지고 있어서 너무 많은 역할을 하고, 컴포즈 화면과 액티비티 로직이 한 파일에 섞여있다고 생각이 들었어요 

그래서 이 파일을 위 커밋과 같이 분리했습니다. 
하지만 XXScreen 란 네이밍을 이미 사용하고 있어서 액티비티와 분리된 컴포저블의 네이밍을 어떻게 해야할지 모르겠어 AI의 도움을 좀 받았는데요 ㅎㅎ 🙃

"화면의 진입점" 이란 의미로 xxRoute란 네이밍을 많이 사용한다고 하더라구요 !
그래서 xxRoute란 네이밍을 사용하였습니다

cucumber99 commented on app/src/main/java/woowacourse/payments/cards/CardScreenUiEvent.kt:
```diff
@@ -0,0 +1,6 @@
+sealed interface CardScreenUiEvent {
+    object Idle : CardScreenUiEvent
+    object CompleteAddCard : CardScreenUiEvent
+}
```
`data object`가 아닌 `object`를 사용한 이유는 무엇인가요?

chanho0908 replied:
음,,,? 그러네요 ...? Android Studio를 9월 1일자 버전으로 업데이트하니 이전 버전에서 나오던 경고가 안보이는군요 ,,, 🧐
이건 제가 놓쳤습니다 😓
<img width="390" height="175" alt="image" src="https://github.com/user-attachments/assets/c261ad95-fa63-4d11-9c9d-841af11f1046" />

cucumber99 commented on app/src/main/java/woowacourse/payments/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,137 @@
+        horizontalAlignment = Alignment.CenterHorizontally
+    ) {
+        when {
+            cards.isEmpty() -> {
```
1. 카드가 없음
2. 카드가 1장만 있음
3. 카드가 2장 이상임

`List<Card>`의 상태라고 생각하는데요.
이 조건을 enum이나 sealed class 혹은 interface로 선언하는 방법은 어떠신가요??
이렇게 한다면 예외 상황도 처리할 수 있을 것 같아요!

chanho0908 replied:
[54e6cf8](https://github.com/woowacourse/android-payments/pull/28/commits/54e6cf816c415027e511889b632e503179bd52f3)

He is Android Gosu 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/component/PaymentsToolbar.kt:
```diff
@@ -0,0 +1,60 @@
+        },
+        actions = {
+            if (addButtonVisible) {
+                TextButton(onClick = { onAddClick() }) {
```
`TextButton(onClick = onAddClick)`
이렇게 해도 되지 않나요?

chanho0908 replied:
https://github.com/woowacourse/android-payments/pull/28/commits/e02ff7a056d5bdb8b4dd944c431801e94a67daa7

He is Kotlin Gosu 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -0,0 +1,32 @@
+    val ownerName: String,
+    val password: String,
+) {
+    fun formatCardNumber(): String {
+        val visibleLength = CARD_NUMBER_MASKING_LENGTH
+        val visiblePart = number.take(visibleLength)
+        val maskedPart = CARD_MASKING_CHAR.repeat(number.length - visibleLength)
+        return (visiblePart + maskedPart)
+            .chunked(CARD_NUMBER_GROUP_SIZE)
+            .joinToString(CARD_SEPARATOR)
+    }
```
카드 번호에 구분자를 추가하는 로직도 도메인 로직인가요?🤔

chanho0908 replied:
[698ab74](https://github.com/woowacourse/android-payments/pull/28/commits/698ab74efc47b6fabb02eabf2e97be76ff7f2115) 

UI 로직에 관련된 상수와 비즈니스 로직과 관련된 상수를 분리했습니다

[c868f81](https://github.com/woowacourse/android-payments/pull/28/commits/c868f81893607b42a6ca374a1e1f086f8dbf1a72)

UI 와 관련된 로직을 컴포저블 파일로 이동했습니다

cucumber99 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    fun formatExpireDate(): String = expireDate.chunked(CARD_EXPIRE_DATE_GROUP_SIZE)
+        .joinToString(CARD_EXPIRE_DATE_SEPARATOR)
```
이 부분도 마찬가지..

cucumber99 commented on app/src/main/java/woowacourse/payments/domain/CardType.kt:
```diff
@@ -0,0 +1,15 @@
+enum class CardType(
+    val parentAlignment: Alignment,
+    val backgroundColor: Color,
+) {
```
`CardType`도 도메인 패키지에 있는데, `androidx`에 대한 의존성이 있네요?
UI 요소를 결정해주는 객체니까 `ui` 패키지에 두는 것이 맞지 않을까요?🤔

chanho0908 replied:
[b4c704d](https://github.com/woowacourse/android-payments/pull/28/commits/b4c704d9e33e7e0beacc224883831c804201c2f9)

전체 패키지를 수정하면서 core 패키지로 이동했습니다 !

cucumber99 commented on app/src/main/java/woowacourse/payments/preview/CardPreviewParameterProvider.kt:
```diff
@@ -0,0 +1,42 @@
+import androidx.compose.ui.tooling.preview.PreviewParameterProvider
+import woowacourse.payments.domain.Card
+
+class OneCardPreviewParameterProvider : PreviewParameterProvider<Card> {
```
오 이런게 있었군요!

cucumber99 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -12,16 +11,18 @@
-    </application>
 
+        <activity
+            android:name=".MainActivity"
```
카드 추가 화면 네이밍도 개선되면 좋을 것 같아요~~

chanho0908 replied:
NewCardActivity 로 변경했슴니다 !

cucumber99 commented on app/src/main/java/woowacourse/payments/cards/CardsActivity.kt:
```diff
@@ -0,0 +1,91 @@
+        enableEdgeToEdge()
+        setContent {
+            AndroidpaymentsTheme {
+                val cards = remember { mutableStateListOf<Card>() }
```
`var cards by remember { mutableStateOf<List<Card>>() }`
이 코드와는 어떤 차이점이 있나요?

chanho0908 replied:
1단계 미션에 mutableState의 프로퍼티 초기화 방법에 대한 레아의 질문에 답변한 [링크](https://github.com/woowacourse/android-payments/pull/5#discussion_r2323789774)입니다 

해당 링크를 보면 `by` 키워드는 Getter/Setter를 위임하고 있어요 ! 

하지만 `mutableStateListOf`의 경우 리스트라 단일 값에 대한 Getter/Setter를 지원하지 않는다고 오류가 발생합니다.
쉽게 말하면 리스트여서 안되는 것 같아요 !
<img width="1010" height="300" alt="image" src="https://github.com/user-attachments/assets/83fb7351-61ca-4cf6-8a69-4b174e0b36db" />

cucumber99 commented on app/src/main/java/woowacourse/payments/cards/CardsActivity.kt:
```diff
@@ -0,0 +1,91 @@
+        setContent {
+            AndroidpaymentsTheme {
+                val cards = remember { mutableStateListOf<Card>() }
+                var uiEvent by remember {
+                    mutableStateOf<Event<CardScreenUiEvent>>(
+                        Event(CardScreenUiEvent.Idle)
+                    )
+                }
```
단일 이벤트를 위해 `Event` 클래스를 선언하고 사용하신 것 같은데, 혹시 의도한대로 잘 동작하나요?
`LiveData`에서만 사용하다가 이렇게 보니까 예상이 잘 안가네요..🥲

chanho0908 replied:
토끼의 [리뷰](https://github.com/woowacourse/android-payments/pull/28#discussion_r2328228130)를 보면 기존 코드에서 구성 변경시 LiveData로 옵저빙중이던 이벤트가 중복해서 발생하는 것과 동일한 문제가 발생했어요
<img width="792" height="725" alt="image" src="https://github.com/user-attachments/assets/d8bd6f35-2a3f-42e0-b3a8-69842d998e13" />

그래서 상태를 다시 복원시키는 로직이 필요했는데 일전 미션에선 배운 `SingleLiveData`가 생각나더라구요 !

하지만 이번 미션에선 상태를 옵저빙 하는 형태가 아니다보니 `SingleLiveData`까지 사용하진 않았습니다 !
SingleLiveData와 Event에 대한 [좋은 글](https://medium.com/prnd/mvvm%EC%9D%98-viewmodel%EC%97%90%EC%84%9C-%EC%9D%B4%EB%B2%A4%ED%8A%B8%EB%A5%BC-%EC%B2%98%EB%A6%AC%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95-6%EA%B0%80%EC%A7%80-31bb183a88ce)이있어 남깁니다 ㅎㅎ

그리고 우려하신 단일 이벤트 처리에 대한것두 잘 작동합니다 !

https://github.com/user-attachments/assets/74184231-1f7a-46cd-8b1b-ec77231cff45

chanho0908 replied:
아까 오이랑 다시 이야기하면서 발견한 내용을 다시 조사해봤어요

기존엔 peekContent 를 사용해 UiEvent를 가져오고 있었어요 하지만 peekContent의 주석에 나와있는 것 처럼
peekContent는 이벤트가 소비 여부에 상관 없이 항상 content를 반환합니다.

```
package woowacourse.payments.ui.core

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 */
open class Event<out T>(
    private val content: T,
) {
    var hasBeenHandled = false
        private set // Allow external read but not write

    /**
     * Returns the content and prevents its use again.
     */
    fun getContentIfNotHandled(): T? =
        if (hasBeenHandled) {
            null
        } else {
            hasBeenHandled = true
            content
        }

    /**
     * Returns the content, even if it's already been handled.
     */
    fun peekContent(): T = content
}
```

그럼에도 불구하고 토스트 메시지가 한 번만 보이는 이유를 추측해봤는데요 !
아마 컴포즈의 리컴포지션 동작 원리 때문일 것으로 예측되어요.

리컴포지션은 Compose가 상태 변경 시에만 재실행하는 작동 방식인데,
지금 Toast를 보여주는 `CompleteAddCard`는 매번 같은 값이 인자로 들어오기 때문에 (싱글톤이어서 !)
리컴포지션이 발생하지 않은 것으로 추측됩니다.🤔

[6e76574](https://github.com/woowacourse/android-payments/pull/28/commits/6e765745c39e8918ffddd51334ffd44c77ead803)

그래서 `Event.getContentIfNotHandled()`로 이벤트가 한 번만 소비되도록 막고,

LaunchedEffect(event)로 key로 전달받은 이벤트가 변경될 때만 리컴포지션이 발생하도록 안전 장치를 만들었습니다 !

cucumber99 commented on app/src/main/java/woowacourse/payments/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,137 @@
+    val context = LocalContext.current
+    val toastMessage = stringResource(R.string.card_list_add_new_card)
+
+    when (uiEvent) {
+        CardScreenUiEvent.CompleteAddCard -> {
+            Toast.makeText(
+                context,
+                toastMessage,
+                Toast.LENGTH_SHORT
+            ).show()
+        }
+
+        CardScreenUiEvent.Idle -> Unit
+    }
```
UI Event를 액티비티가 아닌 하위 컴포저블에서 처리한 페토만의 이유가 궁금해요~

chanho0908 replied:
저도 오이의 생각처럼 토스트나 스낵바, 화면이동 로직은 액티비티에서 처리하는게 맞다고 생각했는데 이번 미션의 요구사항 때문에 하위 컴포저블에서 처리했습니다 !

<img width="425" height="149" alt="image" src="https://github.com/user-attachments/assets/5914e5ec-3bb5-4343-be94-b45cb8368c7e" />

cucumber99 reviewed (changes requested):
안페~ (안녕하세요 페토라는 뜻ㅎ)
2단계도 고생 많으셨어요!
1단계 머지한지 얼마 안된 것 같은데.. 역시 빠르시네요.....

페토 코드 리뷰할 때마다 느끼는 건데, 오히려 제가 배우고 가는 것 같아요👍
충분히 고려할 만한 부분에 대해 리뷰 남겨놨습니다!
문제가 있거나 같이 고민하면 좋을 법한 내용이 있다면 내일부터 찾아오시거나 DM 주세요~
리팩터링 파이팅!!

chanho0908 committed:
chore: 테스트 given, when, then 컨벤션 적용

chanho0908 committed:
chore: 전체 패키지 수정

chanho0908 committed:
feat: 카드 목록 화면 라우트 분리

chanho0908 committed:
feat: 카드 추가 화면 라우팅 화면 분리

chanho0908 committed:
refactor: 툴바 버튼 클릭 리스너 람다 레퍼런스로 변경

chanho0908 committed:
feat: 화면 회전 시 카드 정보 유지 기능 추가

chanho0908 committed:
feat: 카드 화면 상태에 따른 뷰 분리

chanho0908 committed:
fix: 일회성 이벤트가 중복 실행되는 현상 수정

chanho0908 committed:
fix: CardsScreen uiEvent 매개변수 타입 변경

chanho0908 committed:
refactor: 카드 UI 로직과 관련된 상수 UI 로 이동

chanho0908 committed:
refactor: 카드 번호 및 만료일 포맷팅 로직 RegisteredCard 컴포넌트로 이동

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/cards/CardsScreenTest.kt:
```diff
@@ -0,0 +1,143 @@
+        }
+
+        // then
+        composeTestRule
+            .onNodeWithText("새로운 카드를 등록해주세요")
+            .assertDoesNotExist()
```
`assertIsNotDisplayed()`로 테스트해도 될 것 같은데 `assertDoesNotExist()`를 사용한 이유는 무엇인가요?
혹은 아예 테스트하지 않아도 될 것 같다는 생각이 드는군요🤔

chanho0908 replied:
`assertIsNotDisplayed`는 단순히 가시성만을 확인하는 것으로 알고있어요 ! 

visibility가 Gone으로만 되어도 통과하는데,현재 로직은 카드 사이즈에 따라 컴포저블을 아에 렌더링 하지 않는 방식으로 구현했기 때문에 컴포저블의 생성 유무를 확인하기 위해 `assertDoesNotExist`를 사용했어용

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/cards/CardsScreenTest.kt:
```diff
@@ -0,0 +1,143 @@
+                ownerName = "peto",
+                password = ""
+            )
+        )
```
카드 목록 화면 테스트에서 사용하는 `Card` 리스트를 Fixture로 분리해도 될 것 같아요~

chanho0908 replied:
오이는 `Fitxture`에 대해 어떻게 생각하실까요 ? 

저는 레벨 1 둘리의 리뷰에서 요런 [글](https://jojoldu.tistory.com/615)을 추천 받았어요
오이가 제시해주신 것 처럼 자주 사용되는 테스트 더미에 대해선 저도 당연히 `Fitxture` 분리해야 한다고 생각합니다.

다만 지금처럼 특정 상태에 따른 조건부 테스트의 경우, 직접 하드 코딩하는 것이 이 테스트를 이해하는데 더 도움이 된다 생각해요 !☺️

cucumber99 replied:
저도 검증부는 하드코딩되어야 한다는 의견에 적극 동의하는 입장이에요!
위 글에서 설명하는 이유 중 하나는 테스트에 도메인 로직이 포함되는 것이 위험하기 때문인데요.
페토가 만든 `Card` 객체에서는 검증 로직이 포함되어 있지 않고 단순히 값을 나타내기만 하고 있으니,
`Fixture`로 분리해서 테스트 코드를 더 간결하게 작성할 수 있지 않을까? 해서 위와 같은 리뷰를 남겼어요.
페토만의 이유가 충분한 것 같으니 굳이 분리하지 않으셔도 될 것 같아요요!

cucumber99 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -0,0 +1,14 @@
+package woowacourse.payments.domain
+
+data class Card(
+    val number: String,
+    val expireDate: String,
+    val ownerName: String,
+    val password: String,
+) {
+    companion object {
+        val EMPTY = Card("", "", "", "")
+        const val CARD_MAX_LENGTH = 16
+        const val CARD_NUMBER_MASKING_LENGTH = 8
+    }
+}
```
```kotlin
data class Card(
    val number: String = "",
    val expireDate: String = "",
    val ownerName: String = "",
    val password: String = "",
)
```
동반 객체에 빈 `Card`를 두는 것과 위처럼 기본 값을 부여하는 방식 중 어떤 것이 더 낫다고 생각하시나요?
사소한 부분이지만 저는 이 고민을 하다가 페토처럼 동반 객체에 빈 값을 뒀는데..

chanho0908 replied:
저는 개인적으로 가급적 `Default Parameter`를 지양하는 편이에요 !
이유는 `Default Parameter`를 사용함으로써 전달해야 하는 인자를 까먹고 전달하지 않는 실수를 할 수도 있고,
IDE가 누락된 인자를 캐치하지 못하는 불편함도 있어요

또한, 원시값이 아닌 경우, 예를 들어 객체 타입에 `Default Parameter`를 사용하는 경우 다음과 같은 코드를 작성한다고 가정해볼게용

```
Class InputView(
  val inputValidator = InputValidator()
)
```

여기서 `InputValidator`에 파라미터가 하나만 늘어나더라도 `InputView` 에 수정이 발생하고 그 결과 레벨 1,2 동안 열심히 배웠던 SRP를 위배하는 행위라고 생각해요 🧐

이러한 이유들로 저는 `Default Parameter`를 지양하며 동반 객체에 빈 값을 가진 객체를 사용하는 것은 지양하며
동반 객체에 빈 값을 가진 객체를 초기화하는 이유는 `Null Object Patten`에서 착안한 아이디어에요 !

동반 객체에서 인스턴스를 생성하면 Singletone이기 때문에 빈 객체가 필요한 다른 곳에서도 재활용 할 수 있다는 장점이 있는 것 같습니당

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/core/CardType.kt:
```diff
@@ -0,0 +1,15 @@
+    EMPTY(Alignment.Companion.Center, GrayE5),
+    PENDING(Alignment.Companion.CenterStart, Black33),
+    REGISTERED(Alignment.Companion.CenterStart, Black33)
+}
```
`CardType`이라는 이름을 보았을 때 정렬 방식이 타입에 따라 나뉠 것이라고 예상하지 못할 것 같아요.
중복되는 코드를 없애기 위한 것으로 보이는데, 차라리 카드의 상태 별로 컴포저블 함수를 별도로 나누는 것이 가독성 측면에서는 더 좋을 것 같다는 생각이 들어요.

또한 카드 종류를 `EMPTY`, `PENDING`, `RESGISTERED`로 분류하셨는데, 실제 카드 관련 어플처럼 어느 기업의 카드인지에 따라 카드의 UI가 달라진다고 가정할 때 이 `CardType`을 그대로 사용할 수 있을까요?

chanho0908 replied:
네이밍은 저도 지으면서 마땅한 이름이 떠오르지 않았는데 CardType는 진짜 아니였던 것 같네요 😅

> 실제 카드 관련 어플처럼 어느 기업의 카드인지에 따라 카드의 UI가 달라진다고 가정할 때 이 CardType을 그대로 사용할 수 있을까요?

이건 3단계 요구사항에 있으니 현상 유지하고 3단계에서 직접 문제를 겪으며 수정해 나가보겠습니다 !
좋은 의견 고마워요 🥒

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsRoute.kt:
```diff
@@ -0,0 +1,70 @@
+import woowacourse.payments.ui.serialization.SerializationCard
+import woowacourse.payments.ui.view.cards.CardsActivity.Companion.EXTRA_CARD
+
+@Composable
+fun CardsRoute(
+    onAddCardClick: (ManagedActivityResultLauncher<Intent, ActivityResult>) -> Unit
+) {
```
https://github.com/woowacourse/android-payments/pull/28#discussion_r2328626391

`rememberLauncherForActivityResult`를 아직 사용하지 않은 상태여서 이해도가 부족했기 때문에 위와 같은 리뷰를 남겼었네요🥲

> Activiy가 UI + 상태 관리 + ActivityResult 처리까지 다 가지고 있어서 너무 많은 역할을 하고, 컴포즈 화면과 액티비티 로직이 한 파일에 섞여있다고 생각이 들었어요

오히려 저는 액티비티가 화면 이동 및 결과 처리에 대한 책임을 가져야 한다고 생각했어요.
또한 관련 로직을 컴포저블 내부에 두니 컴포저블 함수가 너무 비대해지는 경험을 했기 때문에 더더욱 액티비티에 있어야 한다고 느꼈는데요.
이 부분에 대한 페토의 의견을 조금 더 들어보고 싶네요 ㅎㅎ

> "화면의 진입점" 이란 의미로 xxRoute란 네이밍을 많이 사용한다고 하더라구요 !
그래서 xxRoute란 네이밍을 사용하였습니다

혹시 [이 코드](https://github.com/android/compose-samples/blob/main/JetNews/app/src/main/java/com/example/jetnews/ui/home/HomeRoute.kt) 보셨나요?
제가 봤을 때는 페토가 말했듯이 `xxRoute`가 화면의 진입점 역할, 추가로 어떤 상태에 따른 라우팅을 정의하는 컴포저블 함수라고 생각하는데요.
위 컴포저블 함수가 하는 역할은 제가 이해한 `xxRoute`의 역할과는 다른 것 같아요.
추후 `CardsActivity`에 다양한 화면이 추가된다면 괜찮겠지만, 지금은 굳이 `CardsScreen`과 분리하지 않고 합쳐서 `CardsScreen`이라고 해도 될 것 같아요.

제 잘못된 리뷰때문에 이렇게 코드를 작성하셨으니, 굳이 반영은 안하셔도 돼요👍

chanho0908 replied:
오 !! 안그래도 자료를 찾고 있었는데 아리가또 큐컴버상 ~~ 👍

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsRoute.kt:
```diff
@@ -0,0 +1,70 @@
+                onAddClick = {
+                    onAddCardClick(activityResultLauncher)
+                },
+                addButtonVisible = cards.size > 1
+            )
```
`CardsScreen`과 분리하지 않았다면 `CardsScreenVisibility.of(cards) = CardsScreenVisibility.MULTIPLE` 이런 식으로 개선할 수 있을 것 같아요~

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,164 @@
+                    cardType = CardType.REGISTERED,
+                    onClick = {},
+                    content = {
+                        RegisteredCard(
+                            cards[0],
+                            CARD_NUMBER_GROUP_SIZE,
+                            CARD_NUMBER_SEPARATOR,
+                            CARD_MASKING_CHAR,
+                            CARD_EXPIRE_DATE_GROUP_SIZE,
+                            CARD_EXPIRE_DATE_SEPARATOR
```
카드 정보를 UI 요소로 변환(구분자 추가, 마스킹 등등)해주는 객체를 따로 만드는 것은 어떨까요? 
`RegisteredCard`를 재사용할 수 있게끔 다양한 인자를 전달하니, 오히려 가독성이 떨어지는 느낌이에요.
페로로가 싫어하시는 UiModel 이라거나... ㅋㅋ!

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    Column(
+        modifier = modifier.fillMaxSize(),
+        horizontalAlignment = Alignment.CenterHorizontally
```
Ktlint를 실행하지 않은 흔적이 보이는군요 ㅎㅎ

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsScreenVisibility.kt:
```diff
@@ -0,0 +1,21 @@
+
+import woowacourse.payments.domain.Card
+
+enum class CardsScreenVisibility {
+    EMPTY,
+    SINGLE,
+    MULTIPLE;
```
`sealed class`, `sealed interface`로도 선언할 수 있을 것 같은데 `enum`을 택한 페토의 이유가 궁금해용~

chanho0908 replied:
전 상태가 필요 없거나 각 하위 클래스마다 독립적인 상태를 가야하는 경우가 아니라면 `enum`을 선택한답니다  ☺️

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardActivity.kt:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    private fun moveToCardsActivity(card: Card) {
+        val intent =
+            CardsActivity.newIntent(this, card.toSerializationCard())
+        setResult(RESULT_OK, intent)
+        finish()
+    }
```
`Activity`라는 접미사는 불필요한 것 같아요!

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardRoute.kt:
```diff
@@ -0,0 +1,37 @@
+    onBackClick: () -> Unit,
+    onSaveClick: (Card) -> Unit
+) {
+    var card by rememberSaveable { mutableStateOf(Card.EMPTY.toSerializationCard()) }
+    Scaffold(
+        topBar = {
+            NewCardTopBar(
+                onBackClick = { onBackClick() },
+                onSaveClick = { onSaveClick(card.toDomain()) }
+            )
+        },
+        modifier = Modifier.fillMaxSize()
+    ) { innerPadding ->
+        NewCardScreen(
+            card = card.toDomain(),
+            onCardChange = { card = it.toSerializationCard() },
+            modifier = Modifier.padding(innerPadding)
+        )
+    }
```
빈 값을 가진 `Card` 객체를 생성하고, `SerializationCard`로 변환한 다음 다시 `toDomain()`을 통해 `Card` 객체로 전달하네요?
위와 같이 코드를 작성한 이유가 궁금해요!

chanho0908 replied:
우선 SerializationCard은 오직 **직렬화를 위한 객체** 입니다

그리고 직렬화를 위한 객체를 만든 이유는 XML 시스템에서 `onSaveInstanceState()`에 상태를 저장할 때 Bundle에 담는 시스템의 경우 원시타입이 아닌 객체 타입을 저장하기 위해선 직렬화가 필요한 것 처럼 

`rememberSaveable` 또한 객체를 저장하기 위해선 직렬화가 필요합니다.

그렇다면 이 객체를 저장하기 위해선 두 가지 선택지가 있을 것 같아요
+ 도메인에 Serializable 어노테이션을 추가해 직렬화한다.
+ 도메인과 같지만 Pacelable 어노테이션을 추가한 오직 직렬화만을 위한 객체를 만든다

페로로와의 레벨 인터뷰 과정에서 위 선택지에 대한 질문을 주고받았었는데, 결론적으로 페로로는 두 번째 선택지 처럼 오직 직렬화 만을 위한 객체를 만들것을 제시해주셨어요. 

저에게는 이게 굉장히 충격적이었는데요, Ui에서 상태 저장 또는 Intent 간에 객체 전달을 위해 도메인에 직렬화 기능을 추가하는 것은 결국 UI의 요구사항에 의해 도메인이 변하는 것이다라고 말씀해 주셨습니다.

저는 이 관점에 정말 크게 공감하여 실제로 프로젝트에서 지금과 같이 오직 직렬화를 위한 객체와 도메인을 분리했어요 !

>빈 값을 가진 Card 객체를 생성하고, SerializationCard로 변환한 다음 다시 toDomain()을 통해 Card 객체로 전달하네요?

앞선 내용을 기반으로 오이의 리뷰에 답변을 남기자면,

직렬화 객체는 구성 변경시 상태 저장과 Intent 간의 데이터 교환시에 직렬화를 위한 책임 외에는 절대로 그 어떤 책임과 역할을 갖지 않습니다. 

때문에 여기선 상태를 저장하기 위해 직렬화 객체를 사용했고, UI 에서 사용되는 것은 직렬화 객체의 책임이 아닌 도메인의 책임이기 때문에 UI에 전달할 땐 도메인으로 다시한번 매핑했습니다 !

cucumber99 reviewed (changes requested):
리팩터링도 고생하셨어요!
아직 시간은 많으니까 리뷰 천천히 보시고 반영해주세요~
아까 오프라인으로 이야기한 부분도 반영해주시길 기다리고 있겠습니다👍

chanho0908 committed:
feat: 카드 목록 화면 UiStateHolder 패턴 적용

chanho0908 committed:
feat: 카드 생성 화면 StateHolder 패턴 적용

- `NewCardUiStateHolder`를 추가하여 UI 상태 및 이벤트 처리를 위임합니다.
- `NewCardRoute`를 삭제하고 `NewCardScreen`으로 통합합니다.
- 테스트 코드를 `new` 패키지로 이동합니다.

chanho0908 committed:
refactor: `moveToCardsActivity` 함수명 변경

chanho0908 reviewed (commented):
오랜만임니다 🥒 ~~

오이와 함께했던 레아와의 엔온원을 통해 제가 어떤 방향으로 레벨 4를 임해야할지 정하게 되었어요 
이전에 구두로 협의했던 것과 다르게 도메인을 적용하는 대신 수업에서 배운 것들을 적용해보려 해요 !

그래서 이번에 StateHolder 패턴을 적용해보았습니다 
이 과정에서 구성 변경에도 상태를 유지하기 위해 수업에서 배우지 않은 것들이 나오게 되었는데
혹시 이해가 안되는 코드들이 있다면 언제든 편하게 말씀해주세요 ☺️

그리고 오이에게 한가지 리뷰 방식에 대해 제안하고 싶은게 있슴니다 !

앞으로 남은 미션동안 수업에서 배운 내용을 기반으로 조금 더 딮하게 공부해서 적용해보려해요
그리고 오이가 리뷰하는 과정에서 모르는 것들을 저에게 질문해 주신다면 서로 많은 것들을 얻어갈 수 있을 것 같은데 어떠실까요 ? 🙂

chanho0908 committed:
build: ktlint 플러그인 추가 및 설정

chanho0908 committed:
chore: ktlint 적용

cucumber99 commented on app/src/androidTest/java/woowacourse/payments/cards/CardUiStateHolderTest.kt:
```diff
@@ -0,0 +1,106 @@
+import woowacourse.payments.ui.view.cards.CardUiStateHolder
+import woowacourse.payments.ui.view.cards.CardsUiState
+
+class CardUiStateHolderTest {
+    private val saver: Saver<CardUiStateHolder, Any> = CardUiStateHolder.Saver
+
```
given/when/then 주석이 추가되었으면 좋겠어요~

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/cards/CardsScreen.kt:
```diff
@@ -0,0 +1,246 @@
+    val toastMessage = stringResource(R.string.card_list_add_new_card)
+
+    val event = uiEvent.getContentIfNotHandled()
+    LaunchedEffect(event) {
+        event?.let {
+            when (it) {
+                CardScreenUiEvent.CompleteAddCard -> {
+                    Toast.makeText(context, toastMessage, Toast.LENGTH_SHORT).show()
+                }
+
+                CardScreenUiEvent.Idle -> Unit
+            }
+        }
+    }
```
이벤트를 한 번 발생시키고 나서 다시 `Idle` 상태로 되돌리면 `Event`를 사용하지 않아도 될 것 같아요!
이 부분에 대해서는 어떻게 생각하시나요?

cucumber99 commented on app/src/main/java/woowacourse/payments/ui/view/new/NewCardUiStateHolder.kt:
```diff
@@ -0,0 +1,46 @@
+        _uiState.value = _uiState.value.copy(card = _uiState.value.card.copy(password = password))
+    }
+
+    companion object {
+        val Saver: Saver<NewCardUiStateHolder, SerializationCard> = Saver(
+            save = { holder -> holder.uiState.card.toSerializationCard() },
+            restore = { NewCardUiStateHolder(NewCardUiState(it.toDomain())) }
+        )
+    }
```
👍👍

cucumber99 reviewed (approved):
2단계 고생 많으셨어요~
StateHolder까지 잘 구현하셨네요👍
남은 리뷰도 3단계에서 반영해주세요!
3단계도 파이팅!
